cilium/socket_option_source_address.cc

```diff
@@ -90,13 +91,26 @@ bool SourceAddressSocketOption::setOption(
                 socket.ioHandle().fdDoNotUse(), dst_addr);
       return true;
     }
+
     // Also skip using the original source address if destination is local.
     // Local destinations have a policy configured on their address.
     if (policy_fs_ && policy_fs_->exists(dst_addr)) {
       ENVOY_LOG(trace, "Skipping restore of local address on socket: {} - destination is local {}",
                 socket.ioHandle().fdDoNotUse(), dst_addr);
       return true;
     }
+
+    // Also skip using the original source address in case of N/S L7 loadbalancing if the
+    // destination identity is WORLD.
+    if (policy_fs_ && dest_fs_ && policy_fs_->is_l7lb_ && !original_source_address_ &&
+        policy_fs_->resolvePolicyId(dest_fs_->getDestinationAddress()->ip()) == Cilium::ID::WORLD) {
```

Comment:
This should likely be a bit wider condition, testing if the destination identity has the local scope bit set, as then this would cover WORLDIPv4 and WORLDIPv6, as well as any identities derived from toFQDNs or CIDR policies.